ultravox/data/datasets.py

```diff
@@ -550,77 +549,48 @@ def __len__(self):
         return self._length
 
 
-class StopStrategy(str, enum.Enum):
-    FIRST_EXHAUSTED = "FIRST_EXHAUSTED"
-    LAST_EXHAUSTED = "LAST_EXHAUSTED"
-    NEVER_STOP = "NEVER_STOP"
-
-
 class InterleaveDataset(SizedIterableDataset):
-    """Interleaves multiple IterableDataset objects based on normalized weights."""
+    """Interleaves multiple SizedIterableDataset objects based on normalized weights."""
 
     def __init__(
         self,
         datasets: Sequence[SizedIterableDataset],
         weights: Optional[Sequence[float]] = None,
-        stop_strategy: StopStrategy = StopStrategy.LAST_EXHAUSTED,
-        seed: Optional[int] = 42,
-        static: bool = False,
     ) -> None:
         """
         Args:
             datasets: A list of SizedIterableDataset objects.
-            weights: A list of weights for each dataset.
-            stop_strategy: Strategy for stopping iteration.
+            weights: An optional list of dataset weights, i.e., the number of times it should be repeated.
             seed: Optional seed for reproducibility.
-            static: If true, the datasets are interleaved in a static order with equal weights.
         """
         self._datasets = datasets
-        self._rng = np.random.default_rng(seed)
-        self._static = static
-        self._stop_strategy = stop_strategy
-
-        if weights is None:
+        if weights is not None:
+            assert len(weights) == len(datasets)
+        else:
             weights = [1.0] * len(datasets)
-        total_weight = sum(weights)
-        self._normalized_probs = [w / total_weight for w in weights]
+        self._weighted_samples = [int(w * len(d)) for w, d in zip(weights, datasets)]
+        self._total_samples = sum(self._weighted_samples)
 
     def __iter__(self):
-        # If no datasets are provided, return an empty iterator
-        if not self._datasets:
-            return
-
-        iters = [iter(ds) for ds in self._datasets]
-        exhausted = [False] * len(iters)
-
-        if self._static:
-            static_iter = itertools.cycle(range(len(self._datasets)))
-
-        while True:
-            if self._static:
-                iter_index = next(static_iter)
-            else:
-                iter_index = self._rng.choice(len(iters), p=self._normalized_probs)
-
+        ds_iters = [iter(ds) for ds in self._datasets]
+        ds_pos = [0] * len(ds_iters)
```

Comment:
That probably works in practice, but it breaks the interface (since different people might call `iter` in different places) and if there's a bug it's gonna be hard to catch.

Reply:
Let's take a step back first: what behavior do we think is correct here?

Reply:
Farzad has a fair point. Here are the requirements I have in mind:

1. Already met:
   a) Shuffling of samples within individual datasets.
   b) The number (or distribution) of selected samples from each dataset should align with the specified weight for an epoch (or at any point during training).
2. Not yet met:
   a) Effective use of all samples from a dataset
       Although weight is specified as the percentage of samples used for training in an epoch, it actually reflects the relative importance of the datasets. Consider datasets A and B, each with 1M samples, where A is noisier than B. We may assign weights of 0.1 to A and 1.0 to B, and train for 3 epochs. In this case, we definitely want a different subset of samples from A to be used in the second epoch, and similarly for the third.
   b) New shuffling of samples when a dataset is iterated in subsequent epochs of the dataset (not the specified training epochs)


Regarding the difference between deterministic and random sample, it seems to me that the main advantage of the deterministic approach is to make sure that the exact number of samples from each dataset is used in an epoch, and the distribution of selected samples matches the specified distribution as close as possible at any point of training. These are nice attributes to have.